File: 103.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (!root) return {};
        vector<vector<int>> res;
        queue<TreeNode*> q{{root}};
        // 标记位，奇数位置需要翻转，从 0 开始，第一个是从左到右
        bool leftToRight = true;
        while (!q.empty()) {
            // 通过赋值就不会导致 q.size() 的变化而影响到 i 的遍历区间了
            int size = q.size();
            vector<int> oneLevel(size);
            for (int i = 0; i < size; ++i) {
                TreeNode *t = q.front(); q.pop();
                // 如果需要反向输出则 idx = size - 1 - i;
                int idx = leftToRight ? i : (size - 1 - i);
                oneLevel[idx] = t->val;
                if (t->left) q.push(t->left);
                if (t->right) q.push(t->right);
            }
            // 每输出一行自动取反一次
            leftToRight = !leftToRight;
            res.push_back(oneLevel);
        }
        return res;
    }
};

File: Medium/Tree/103.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def zigzagLevelOrder(self, root:TreeNode):
        if not root:
            return []
        result = []
        queue = deque([root])

        left_to_right = True

        while queue:
            level_size = len(queue)
            level_nodes = []

            for _ in range(level_size):
                node = queue.popleft()
                level_nodes.append(node.val)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            if not left_to_right:
                level_nodes.reverse()

            result.append(level_nodes)

            left_to_right = not left_to_right
        
        return result
    
if __name__ == "__main__":
    # construct binary tree
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    root.right.right = TreeNode(6)
    s = Solution()
    result = s.zigzagLevelOrder(root)

    print("Zig Zag result:")
    for i, level in enumerate(result):
        direction = "left to right" if i % 2 == 0 else "right to left"
        print(f"Level {i + 1} {direction}: {level}")
    
